### PR TITLE
add http2-debug-state feature

### PR DESCRIPTION
**Pull request by i110**
Implemented `HTTP/2 Implementation Debug State` which is described below:
https://tools.ietf.org/html/draft-benfield-http2-debug-state-00

Sample Response:
```
< HTTP/2 200
< server: h2o/2.1.0-DEV
< date: Tue, 09 Aug 2016 09:50:17 GMT
< content-type: application/json; charset=utf-8
< cache-control: no-cache, no-store
< conn-flow-in: 65535
< conn-flow-out: 65535
<
{
  "settings": {
    "SETTINGS_HEADER_TABLE_SIZE": 4096,
    "SETTINGS_ENABLE_PUSH": 0,
    "SETTINGS_MAX_CONCURRENT_STREAMS": 100,
    "SETTINGS_INITIAL_WINDOW_SIZE": 16777216,
    "SETTINGS_MAX_FRAME_SIZE": 16384
  },
  "peerSettings": {
    "SETTINGS_HEADER_TABLE_SIZE": 4096,
    "SETTINGS_ENABLE_PUSH": 1,
    "SETTINGS_MAX_CONCURRENT_STREAMS": 4294967295,
    "SETTINGS_INITIAL_WINDOW_SIZE": 65535,
    "SETTINGS_MAX_FRAME_SIZE": 16384
  },
  "connFlowOut": 65535,
  "connFlowIn": 65535,
  "streams": {
    "1": {
      "state": "OPEN",
      "flowIn": 16777216,
      "flowOut": 65535,
      "dataIn": 0,
      "dataOut": 0
    }
  },
  "hpack": {
    "inbound_table_size": 3,
    "inbound_dynamic_header_table": [
      [ "accept", "*/*" ],
      [ "user-agent", "curl/7.49.1" ],
      [ ":authority", "127.0.0.1:8080" ]
    ],
    "outbound_table_size": 0,
    "outbound_dynamic_header_table": [
    ]
  },
  "sentGoAway": false
}
```

Points to consider:

* Using debugging URI with proxies may have security issues (being discussed at https://github.com/python-hyper/draft-http2-debug-state/issues/7)
    * When a connection sharing happens in a proxy, one's hpack state (including security-concerned fields) may be exposed to another
    * Perhaps the next draft will mention a switch which can be used to specify whether `hpack` property is included in the response or not (see: https://github.com/python-hyper/draft-http2-debug-state/commit/532c5315c8b8ab28993eff69e232dd261096359d)
    * This implementation doesn't have such a switch yet, because I understand and assume that in almost cases this feature will be used in the controllable environment, but we should watch and see the discussion, and add that switch when the next draft is published
* When a request comes via HTTP/1, I guess that the server should resnpond with 404
    * https://github.com/python-hyper/draft-http2-debug-state/issues/7

### CODE REVIEW

File: lib/handler/status.c [hunk 1]
```diff
@@ -112,7 +112,7 @@ static void send_response(struct st_h2o_status_collector_t *collector)
     resp[cur_resp++] = (h2o_iovec_t){H2O_STRLIT("\n}\n")};
 
     req->res.status = 200;
-    h2o_add_header(&req->pool, &req->res.headers, H2O_TOKEN_CONTENT_TYPE, H2O_STRLIT("text/plain; charset=utf-8"));
+    h2o_add_header(&req->pool, &req->res.headers, H2O_TOKEN_CONTENT_TYPE, H2O_STRLIT("application/json; charset=utf-8"));
```
Comment (i110) [inline lib/handler/status.c]: @kazuho I thought this is better, WDYT?

File: include/h2o/http2_internal.h [hunk 2]
```diff
@@ -50,6 +50,12 @@ typedef struct st_h2o_hpack_header_table_t {
     size_t hpack_max_capacity; /* the value set by SETTINGS_HEADER_TABLE_SIZE */
 } h2o_hpack_header_table_t;
 
+typedef struct st_h2o_hpack_header_table_entry_t {
+    h2o_iovec_t *name;
+    h2o_iovec_t *value;
+    const char *err_desc; /* the recorded soft error description */
+} h2o_hpack_header_table_entry_t;
```
Comment (i110) [inline include/h2o/http2_internal.h]: moved from `hpack.c` to expose it so that debug_state handler can use this..

File: lib/core/token_table.h [hunk 3]
```diff
@@ -21,68 +21,72 @@
  */
 
 /* DO NOT EDIT! generated by tokens.pl */
-h2o_token_t h2o__tokens[] = {{{H2O_STRLIT(":authority")}, 1, 0, 0, 0, 0},
```
Comment (i110) [inline lib/core/token_table.h]: doesn't change except for adding `conn-flow-in` and `conn-flow-out`, but re-formatted by `tokens.pl`

File: lib/handler/status.c (hunk 1 above)
Comment (kazuho) [inline lib/handler/status.c]: Is this change relevant to the PR?
(and I believe we should use `text/plain` considering the fact that it can be displayed using web browsers...)

File: lib/http2/hpack.c [hunk 4]
```diff
@@ -264,6 +258,11 @@ static inline struct st_h2o_hpack_header_table_entry_t *header_table_get(h2o_hpa
     return entry;
 }
 
+inline h2o_hpack_header_table_entry_t *h2o_hpack_header_table_get(h2o_hpack_header_table_t *header_table, size_t index)
+{
+    return (h2o_hpack_header_table_entry_t*)(header_table_get(header_table, index));
+}
+
```
Comment (kazuho) [inline lib/http2/hpack.c]: All `inline` functions must be also defined as `static`.

Also, I would suggest refraining from creating bridges (i.e. a function that just calls another function), since it makes the code-base unnecessary complex. Our general preference goes to keeping the code-base compact and precise, rather than minimizing impact of each PR.

In this case, I would suggest changing `static inline struct st_h2o_hpack_header_table_entry_t *header_table_get(...)` to `h2o_hpack_eader_table_entry_t *h2o_hpack_header_table_get(...)`. Even without `inline` attribute, compilers are allowed to (and will) inline-expand functions defined in the same source file as the caller.

**Comment by kazuho** [comment]
Thank you for the PR! It's good to have the debugging feature on H2O.

Please take a look at my in-line comments as well as two high-level concerns stated below regarding the implementation.

Let’s not make `conn-flow-in` and `conn-flow-out` headers as tokens, considering the fact that adding a token has _some_ impact on performance (due to the increased lookup cost), and that the two headers would never be used in a performance-sensitive way.

How about defining the JSON generator as a connection-level callback (i.e. as one of the various h2-only callbacks in `h2o_conn_callbacks_t`. That way, there would be no need to expose the internals of HTTP/2 implementation to the handler.

File: lib/handler/debug_state.c [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+         conn->peer_settings.enable_push,
+         conn->peer_settings.max_concurrent_streams,
+         conn->peer_settings.initial_window_size,
+         conn->peer_settings.max_frame_size,
+         conn_flow_out.base,
+         conn_flow_in.base);
+
+    /* encode streams */
+    {
+        int coma_removed = 0;
```
Comment (kazuho) [inline lib/handler/debug_state.c]: s/coma/comma/g?

File: lib/handler/debug_state.c [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+                   "      \"flowOut\": %zd,\n"
+                   "      \"dataIn\": %zu,\n"
+                   "      \"dataOut\": %zu\n"
+                   "    }",
+                   stream->stream_id,
+                   state_string,
+                   stream->input_window._avail,
+                   stream->output_window._avail,
+                   (stream->_req_body == NULL ? 0 : stream->_req_body->size),
+                   stream->req.bytes_sent);
```
Comment (kazuho) [inline lib/handler/debug_state.c]: I would appreciate it if we could use memory less aggressively. In the current implementation, `OUTPUT` allocates 2KB chunk, and that means that we would consume at max 2*(100+100)=400KB of memory here.

File: lib/handler/debug_state.c [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+    do {                                                                                         \
+        h2o_hpack_header_table_t *header_table = ht;                                   \
+        int i, coma_removed = 0;                                                                    \
+        for (i = 0; i < header_table->num_entries; i++) {                                    \
+            h2o_hpack_header_table_entry_t *entry = h2o_hpack_header_table_get(header_table, i); \
+            OUTPUT(                                                                              \
+                   ",\n"                                                                         \
+                   "      [ \"%s\", \"%s\" ]",                                                    \
+                   entry->name->base,                                                            \
+                   entry->value->base);                                                          \
```
Comment (kazuho) [inline lib/handler/debug_state.c]: Please use `%.*s` and specify the lengths of the values using `->len` for both `name` and `value`.

Also, please consider the case in which the name (or value) is very long. Current approach of allocating a fixed amount of memory would not work well.

File: lib/handler/status.c (hunk 1 above)
Comment (i110) [inline lib/handler/status.c]: Sorry, this isn't relevant to this PR.
I believe that canonical MIME type of JSON document is `application/json`, and also this json document are consumed by only "status" web-app, so people don't see in browsers, isn't it?
https://tools.ietf.org/html/rfc7159#section-11

But, I have no strong kodawari about this, so can get it back to`text/plain` ( also with `debug_state.c` )

File: lib/handler/debug_state.c (hunk 5 above)
Comment (i110) [inline lib/handler/debug_state.c]: I found this spelling in `lib/handler/status.c`, and misunderstood it as a oshare-way of comma.
I'll fix this!

File: lib/http2/hpack.c (hunk 4 above)
Comment (i110) [inline lib/http2/hpack.c]: > keeping the code-base compact and precise, rather than minimizing impact of each PR
totally agree.
I'll fix.

File: lib/handler/debug_state.c (hunk 7 above)
Comment (i110) [inline lib/handler/debug_state.c]: tashikani..

File: lib/handler/debug_state.c (hunk 6 above)
Comment (i110) [inline lib/handler/debug_state.c]: OK, I'll reduce memory footprint

**Comment by i110** [comment]
@kazuho Thank you for your comments, I'll fix these problems,
BTW, new draft was published yesterday:
https://tools.ietf.org/html/draft-benfield-http2-debug-state-01
I'll follow this new draft from now.

**Comment by kazuho** [comment]
Regarding memory footprint, since this is not a performance critical patch, I would suggest calling `snprintf` repeatedly until it succeeds in buiding the entire string, every time increasing the allocated buffer size by 2x (i.e. exponential back-off).

By using exponential back-off, you can guarantee that the size of memory consumed would be below the actual size of the output multiplied by a constant factor (which would be 3 in this case).

Note: from `man snprintf` on Mac OS X:
```
    The snprintf() and vsnprintf() functions will write at most size-1 of the characters printed into the output string
     (the size'th character then gets the terminating `\0'); if the return value is greater than or equal to the size
     argument, the string was too short and some of the printed characters were discarded.  The output is always null-ter-
     minated, unless size is 0.
```

**Comment by kazuho** [comment]
Oops. Updated my previous comment. Exponential back-off isn't required in this case.

**Comment by i110** [comment]
@kazuho fixed issues.
Also cached up on draft-01. But some new added fields (`queuedData`, `paddingIn`, `paddingOut` ) are ommited because these are optional, and to add these fields we have to track additional states of streams, which I concerned about performance effect on.

File: lib/handler/debug_state.c [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+        req->res.status = 404;
+        h2o_add_header(&req->pool, &req->res.headers, H2O_TOKEN_CONTENT_TYPE, H2O_STRLIT("text/plain; charset=utf-8"));
+        h2o_add_header(&req->pool, &req->res.headers, H2O_TOKEN_CACHE_CONTROL, H2O_STRLIT("no-cache, no-store"));
+        h2o_start_response(req, &generator);
+        h2o_send(req, NULL, 0, 1);
+        return 0;
+    }
+
+    const h2o_token_t conn_flow_in_token = { { H2O_STRLIT("conn-flow-in") } };
+    const h2o_token_t conn_flow_out_token = { { H2O_STRLIT("conn-flow-out") } };
```
Comment (kazuho) [inline lib/handler/debug_state.c]: You cannot do this.

`h2o_token_t` is an object that be compared by checking their address. To achieve such property, a string contained by a token MUST be interned (i.e. only one `h2o_token_t` should exist within the entire program for each string). Trying to define a token in each source file collides with the requirement.

You should instead use `h2o_add_header_by_str` for registering a header that is (either guaranteed or might) not be a token.

File: include/h2o/http2.h [hunk 9]
```diff
@@ -60,6 +60,15 @@ extern const h2o_http2_priority_t h2o_http2_default_priority;
 void h2o_http2_accept(h2o_accept_ctx_t *ctx, h2o_socket_t *sock, struct timeval connected_at);
 int h2o_http2_handle_upgrade(h2o_req_t *req, struct timeval connected_at);
 
+/* debug state */
+typedef struct st_h2o_http2_debug_state_t{
+    h2o_iovec_vector_t json;
+    ssize_t conn_flow_in;
+    ssize_t conn_flow_out;
```
Comment (kazuho) [inline include/h2o/http2.h]: Use of `size_t` is preferred unless you need to use negative values to represent error conditions.

File: include/h2o/http2_internal.h [hunk 10]
```diff
@@ -224,6 +232,7 @@ struct st_h2o_http2_conn_t {
     } _write;
     h2o_cache_t *push_memo;
     h2o_http2_casper_t *casper;
+    int _sent_goaway;
```
Comment (kazuho) [inline include/h2o/http2_internal.h]: I believe `conn->state == H2O_HTTP2_CONN_STATE_HALF_CLOSED` would be sufficient to see if a GOAWAY frame has been sent. Please refer to the comments on `h2o_http2_conn_state_t`.

File: lib/http2/debug_state.c [hunk 11]
```diff
@@ -0,0 +1,210 @@
...
+    char check[1]; /* for only checking the size of output */
+
+#define OUTPUT(_fmt, ...)                                        \
+    do {                                                         \
+        int size = snprintf(check, 1, _fmt, __VA_ARGS__);        \
+        h2o_iovec_t v;                                           \
+        v.base = h2o_mem_alloc_pool(&req->pool, size + 1);             \
+        v.len = snprintf(v.base, size + 1, _fmt, __VA_ARGS__);   \
+        state->json.entries[state->json.size++] = v; \
+    } while (0)
```
Comment (kazuho) [inline lib/http2/debug_state.c]: I'd suggest making this a function (instead of macro), considering the fact that the arguments are evaluated more than once, and because this is not performance-critical.

File: lib/http2/debug_state.c [hunk 12]
```diff
@@ -0,0 +1,210 @@
...
+                   ",\n"                                                                         \
+                   "      [ \"%.*s\", \"%.*s\" ]",                                               \
+                   (int)entry->name->len, entry->name->base,                                     \
+                   (int)entry->value->len, entry->value->base);                                  \
+            if (state->json.entries[state->json.size - 1].len > 0 && !comma_removed) {           \
+                state->json.entries[state->json.size - 1].base[0] = ' ';                         \
+                comma_removed = 1;                                                               \
+            }                                                                                    \
+        }                                                                                        \
+    } while (0)
```
Comment (kazuho) [inline lib/http2/debug_state.c]: Let's also make this a function, so that it would be easier to maintain.

File: lib/http2/debug_state.c [hunk 13]
```diff
@@ -0,0 +1,210 @@
...
+        case H2O_HTTP2_STREAM_STATE_RECV_BODY:
+        case H2O_HTTP2_STREAM_STATE_REQ_PENDING:
+        case H2O_HTTP2_STREAM_STATE_SEND_HEADERS:
+        case H2O_HTTP2_STREAM_STATE_SEND_BODY:
+        case H2O_HTTP2_STREAM_STATE_SEND_BODY_IS_FINAL:
+            return debug_state_string_open;
+        case H2O_HTTP2_STREAM_STATE_IDLE:
+        case H2O_HTTP2_STREAM_STATE_END_STREAM:
+            return NULL;
+        }
```
Comment (kazuho) [inline lib/http2/debug_state.c]: State of a pulled stream changes to `half-closed (remote)` as soon as a HEADER or CONTINUATION frame with the `END_STREAM` flag being set is received by the server.

**Comment by kazuho** [comment]
@i110 Thank you for the changes as well as following the updates to the I-D!

> * I couldn't found the benefit of using connection callbacks for JSON generation, so made the handler directly call `h2o_get_http2_debug_state` 

`if (req->version < 0x200) return NULL;` is not a type-safe way to check _which_ protocol implementation is used. Consider the case when HTTP/3 is introduced. Would this code work as expected? Using callback slots that gets initialized by each protocol handler does not have this issue (in other words, use of callbacks is type-safe).

And the added bonus would be that you can remove the tight coupling between lib/handler/debug_state.c and h2o/http2.h.

> * `s/coma/comma/g` also in `status.c` (If you hope another pr, I'll create it)
> * `application/json` についてはどうしますかね。 Should I restore it to `text/plain` ? https://github.com/h2o/h2o/pull/1019#discussion_r74546907

You should always create a separate PR for separate issues.

Consider the case when we find a bug related to this PR. If this PR was dedicated to the issue, then we can simply run `git revert`. Otherwise, we would need to look into each line of the code to see how to do it.

**Comment by i110** [comment]
@kazuho Thank you for your comments.
I fixed the issues you said, please take a look at commit comments.
Additionaly, as we talked, I removed `sentGoAway` field because of state tracking issue (and of cause this is optional)

File: include/h2o.h [hunk 14]
```diff
@@ -673,6 +673,18 @@ typedef struct st_h2o_res_t {
...
+/**
+ * debug state (currently only for HTTP/2)
+ */
+typedef struct st_h2o_debug_state_t{
+    h2o_iovec_vector_t json;
+    ssize_t conn_flow_in;
+    ssize_t conn_flow_out;
+} h2o_debug_state_t;
+
+h2o_debug_state_t *h2o_http2_get_debug_state(h2o_req_t *req, int hpack_enabled);
```
Comment (kazuho) [inline include/h2o.h]: Please move the prototype to http2_internal.h, since the function is going to be referred via callback from outside of the http2 stack.

File: include/h2o.h [hunk 15]
```diff
@@ -673,6 +673,18 @@ typedef struct st_h2o_res_t {
...
 
+
+/**
+ * debug state (currently only for HTTP/2)
+ */
+typedef struct st_h2o_debug_state_t{
+    h2o_iovec_vector_t json;
+    ssize_t conn_flow_in;
+    ssize_t conn_flow_out;
+} h2o_debug_state_t;
```
Comment (kazuho) [inline include/h2o.h]: Please rename the struct back to `h2o_http2_debug_state_t`, since the structure is dependent to HTTP/2 (as per defined by the I-D, right?).

File: lib/core/token_table.h [hunk 16]
```diff
@@ -21,67 +21,69 @@
...
+    { { H2O_STRLIT("transfer-encoding") }, 57, 1, 1, 1, 0 },
+    { { H2O_STRLIT("upgrade") }, 0, 1, 1, 1, 0 },
+    { { H2O_STRLIT("user-agent") }, 58, 0, 0, 0, 1 },
+    { { H2O_STRLIT("vary") }, 59, 0, 0, 0, 0 },
+    { { H2O_STRLIT("via") }, 60, 0, 0, 0, 0 },
+    { { H2O_STRLIT("www-authenticate") }, 61, 0, 0, 0, 0 },
+    { { H2O_STRLIT("x-forwarded-for") }, 0, 0, 0, 0, 0 },
+    { { H2O_STRLIT("x-reproxy-url") }, 0, 0, 0, 0, 0 },
+    { { H2O_STRLIT("x-traffic") }, 0, 0, 0, 0, 0 }
+};
```
Comment (kazuho) [inline lib/core/token_table.h]: Please completely revert the changes made to lib/core/token_table.h, so that the file would be left unchanged upon PR merger.

File: lib/handler/configurator/debug_state.c [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+                return -1;
+            }
+        }
+        h2o_debug_state_register(ctx->hostconf, hpack_enabled);
+        return 0;
+    } break;
+    default:
+        h2o_configurator_errprintf(cmd, node, "node must be a scalar or a mapping");
+        return -1;
+    }
```
Comment (kazuho) [inline lib/handler/configurator/debug_state.c]: How about defining this as a configuration directive that takes a scalar argument of `minimum`, `hpack`? Doing that way, we can eliminate the complexity of handling a mapping.

The handler would be _off_ in case the the configuration directive is not used.

File: lib/handler/debug_state.c [hunk 18]
```diff
@@ -0,0 +1,72 @@
...
+    static h2o_generator_t generator = {NULL, NULL};
+
+    if (req->conn->callbacks->get_debug_state == NULL) {
+        req->res.status = 404;
+        h2o_add_header(&req->pool, &req->res.headers, H2O_TOKEN_CONTENT_TYPE, H2O_STRLIT("text/plain; charset=utf-8"));
+        h2o_add_header(&req->pool, &req->res.headers, H2O_TOKEN_CACHE_CONTROL, H2O_STRLIT("no-cache, no-store"));
+        h2o_start_response(req, &generator);
+        h2o_send(req, NULL, 0, 1);
+        return 0;
+    }
```
Comment (kazuho) [inline lib/handler/debug_state.c]: Simply returning -1 (so that the next generator would handle the request) might be sufficient.

File: lib/http2/debug_state.c [hunk 19]
```diff
@@ -0,0 +1,217 @@
...
+            return debug_state_string_half_closed_remote;
+        case H2O_HTTP2_STREAM_STATE_IDLE:
+        case H2O_HTTP2_STREAM_STATE_END_STREAM:
+            return NULL;
+        }
+    }
+    return NULL;
+}
+
+static void append_line(h2o_mem_pool_t *pool, h2o_iovec_vector_t *lines, const char *fmt, ...)
```
Comment (kazuho) [inline lib/http2/debug_state.c]: Please use ` __attribute__((format(printf, xxx, yyy)))` extension to detect errors in format specifiers. Examples can be found in the headers of H2O, and I assume documentation can be found on gcc.gnu.org.

File: lib/http2/debug_state.c [hunk 20]
```diff
@@ -0,0 +1,217 @@
...
+    va_list args;
+
+    va_start(args, fmt);
+    int size = vsnprintf(check, 1, fmt, args);
+    va_end(args);
+
+    h2o_iovec_t v;
+    v.base = h2o_mem_alloc_pool(pool, size + 1);
+
+    va_start(args, fmt);
```
Comment (kazuho) [inline lib/http2/debug_state.c]: Please use `va_copy`. Applying `va_start` more than once against one `va_list` is not permitted (IIRC the behavior is undefined).

File: lib/http2/debug_state.c [hunk 21]
```diff
@@ -0,0 +1,217 @@
...
+    state->conn_flow_in = conn->_write.window._avail;
+    state->conn_flow_out = conn->_write.window._avail;
+
+    /* calculate total number of JSON lines */
+    size_t nr_resp = 3;
+    nr_resp += conn->num_streams.pull.open + conn->num_streams.push.open; // streams
+    if (hpack_enabled) {
+        nr_resp += 3 + conn->_input_header_table.num_entries + conn->_output_header_table.num_entries; // hpack
+    }
+    h2o_vector_reserve(&req->pool, &state->json, nr_resp);
```
Comment (kazuho) [inline lib/http2/debug_state.c]: Let's call h2o_vector_reserve in `append_line` instead of trying to determine the correct length of the output here.

This is not a performance sensitive handler, and minimizing the risks of having issues should be given higher priority than optimization.

**Comment by kazuho** [comment]
@i110 Thank you for the updates.

I like the code. It has become even more precise and simple, it's getting near to merger.

Please take a look at my comments in-line.

File: lib/handler/configurator/debug_state.c (hunk 17 above)
Comment (i110) [inline lib/handler/configurator/debug_state.c]: great idea!

File: lib/handler/debug_state.c (hunk 18 above)
Comment (i110) [inline lib/handler/debug_state.c]: tashikani.

File: lib/http2/debug_state.c (hunk 20 above)
Comment (i110) [inline lib/http2/debug_state.c]: memo:
As we talked, `va_start` can be called again after `va_end` is called against the same va_list

**Comment by i110** [comment]
@kazuho: issues fixed :smile:

File: lib/http2/debug_state.c [hunk 22]
```diff
@@ -0,0 +1,210 @@
...
+#include <inttypes.h>
+#include "h2o.h"
+#include "h2o/http2.h"
+#include "h2o/http2_internal.h"
+
+static const char *debug_state_string_open = "OPEN";
+static const char *debug_state_string_half_closed_remote = "HALF_CLOSED_REMOTE";
+static const char *debug_state_string_reserved_local = "RESERVED_LOCAL";
+
+const char *get_debug_state_string(h2o_http2_stream_t *stream)
```
Comment (kazuho) [inline lib/http2/debug_state.c]: maybe add `static`?

File: lib/http2/debug_state.c [hunk 23]
```diff
@@ -0,0 +1,210 @@
...
+            return debug_state_string_half_closed_remote;
+        case H2O_HTTP2_STREAM_STATE_IDLE:
+        case H2O_HTTP2_STREAM_STATE_END_STREAM:
+            return NULL;
+        }
+    }
+    return NULL;
+}
+
+__attribute__((format(printf, 3, 4))) static void append_line(h2o_mem_pool_t *pool, h2o_iovec_vector_t *lines, const char *fmt, ...)
```
Comment (kazuho) [inline lib/http2/debug_state.c]: Renaming this function to `append_block` or `append_chunk` might be a good idea considering the fact that this function is never used to append a _line_.

**Comment by kazuho** [comment]
Thank you for adding `http2`-prefix to the source file. Could you please do the same for the name of the configuration directive and for the global function exported by the handler?

File: lib/handler/http2_debug_state.c [hunk 24]
```diff
@@ -0,0 +1,67 @@
...
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+ * IN THE SOFTWARE.
+ */
+#include "h2o.h"
+
+#include <inttypes.h>
```
Comment (kazuho) [inline lib/handler/http2_debug_state.c]: Please include system headers first and then project-specific headers.

This is necessary since we _might_ want to override some system-defined headers.

File: lib/handler/http2_debug_state.c [hunk 25]
```diff
@@ -19,18 +19,17 @@
...
  */
-#include "h2o.h"
-
 #include <inttypes.h>
+#include "h2o.h"
 
-struct st_h2o_debug_state_handler_t {
+struct st_h2o_http2_debug_state_handler_t {
     h2o_handler_t super;
     int hpack_enabled : 1;
```
Comment (kazuho) [inline lib/handler/http2_debug_state.c]: If you want a boolean, please use `unsigned : 1` that takes either 0 or 1. `int :1` takes two either 0 or -1.

Or simply using `int` or `unsigned` without `:nnn` would be sufficient in this case, considering the fact that the size of the struct would be indifferent and that we wouldn't have many instances of this struct.

File: lib/http2/http2_debug_state.c [hunk 26]
```diff
@@ -0,0 +1,212 @@
...
+{
+    int i, comma_removed = 0;
+    for (i = 0; i < header_table->num_entries; i++) {
+        h2o_hpack_header_table_entry_t *entry = h2o_hpack_header_table_get(header_table, i);
+        append_chunk(pool, chunks,
+                    ",\n"
+                    "      [ \"%.*s\", \"%.*s\" ]",
+                    (int)entry->name->len, entry->name->base,
+                    (int)entry->value->len, entry->value->base);
+        if (!comma_removed) {
```
Comment (kazuho) [inline lib/http2/http2_debug_state.c]: I assume you can change `!comma_removed` to `i == 0`?

File: lib/http2/http2_debug_state.c [hunk 27]
```diff
@@ -0,0 +1,212 @@
...
+            chunks->entries[chunks->size - 1].base[0] = ' ';
```
Comment (kazuho) [inline lib/http2/http2_debug_state.c]: It might be preferable to do the following instead so as not to leave trailing spaces.
```
chunks->entries[chunk->size - 1].base += 1;
chunk->entries[chunk->size - 1].len -= 1;
```

It _might_ be better to define the logic as a separate function, since you would want to do the same at other parts of the code as well.

Comment (kazuho) [inline lib/http2/http2_debug_state.c]: P.S. the other method you can use is to have comma appended to every element and then remove the last comma, or rewrite the first comma to { or [.

File: lib/http2/http2_debug_state.c (hunk 26 above)
Comment (i110) [inline lib/http2/http2_debug_state.c]: you're right..

File: lib/http2/hpack.c [hunk 28]
```diff
@@ -256,7 +250,7 @@ static h2o_iovec_t *decode_string(h2o_mem_pool_t *pool, const uint8_t **src, con
     return ret;
 }
 
-static inline struct st_h2o_hpack_header_table_entry_t *header_table_get(h2o_hpack_header_table_t *table, size_t index)
+h2o_hpack_header_table_entry_t *h2o_hpack_header_table_get(h2o_hpack_header_table_t *table, size_t index)
```
Comment (kazuho) [inline lib/http2/hpack.c]: Assuming that all the structs used inside the function is defined in the headers, would it be possible to change this function to a static inline function declared and defined in http2_internal.h?

Please refer to other inline functions defined in http2_internal.h to how to define a global function as inline (in essence, declare as `static` and, define as `inline`).

File: lib/http2/http2_debug_state.c [hunk 29]
```diff
@@ -86,19 +86,19 @@ __attribute__((format(printf, 3, 4))) static void append_chunk(h2o_mem_pool_t *p
...
                     (int)entry->value->len, entry->value->base);
-        if (!comma_removed) {
-            chunks->entries[chunks->size - 1].base[0] = ' ';
-            comma_removed = 1;
-        }
     }
+
+    if (i > 0)
+        // remove the last commna
+        --chunks->entries[chunks->size - 1].len;
```
Comment (kazuho) [inline lib/http2/http2_debug_state.c]: Please do not insert a comment taking a line within a if statement (without using a block), since it is counterintuitive.

You should write the comment either _above_ the `if` keyword or surround the comment and the statement that is executed if `if` evaluates to true using `{}`.

**Comment by i110** [comment]
@kazuho fixed and applied clang-format.

File: include/h2o.h [hunk 30]
```diff
@@ -1722,6 +1735,17 @@ void h2o_status_register(h2o_pathconf_t *pathconf);
  */
 void h2o_status_register_configurator(h2o_globalconf_t *conf);
 
+/* lib/handler/status.c */
```
Comment (kazuho) [inline include/h2o.h]: wrong filename? :-)

File: lib/http2/http2_debug_state.c [hunk 31]
```diff
@@ -0,0 +1,195 @@
...
+        case H2O_HTTP2_STREAM_STATE_END_STREAM:
+            return NULL;
+        }
+    }
+    return NULL;
+}
+
+__attribute__((format(printf, 3, 4))) static void append_chunk(h2o_mem_pool_t *pool, h2o_iovec_vector_t *chunks, const char *fmt,
+                                                               ...)
+
```
Comment (kazuho) [inline lib/http2/http2_debug_state.c]: unnecessary blank line here?

**Comment by kazuho** [comment]
Going through the code I have found tiny tiny issues. Aside from that all look good to me.

File: include/h2o.h (hunk 30 above)
Comment (i110) [inline include/h2o.h]: ぐぬぬ

**Comment by i110** [comment]
@kazuho fixed!

**Comment by kazuho** [comment]
Thank you for your hard work!
